tools/query_breakdown/src/main/java/com/google/bigquery/QueryBreakdown.java

```diff
@@ -16,52 +15,109 @@
+   * times and create multiple instances of it.
    */
-  public static void run(String originalQuery, String outputFile, int errorLimit) {
+  public QueryBreakdown(Parser parser) {
+    minimumUnparseableComp = Integer.MAX_VALUE;
+    root = new Node();
+    this.parser = parser;
+  }
 
-    // determines which parser to use
-    parser = new CalciteParser();
+  /**
+   * This is the method that will run QueryBreakdown given an original query and output
+   * it to the specified output file or commandline. The provided errorLimit will stop the
+   * tool from running over a certain time.
+   *
+   * TODO: output file feature and runtime limit support
+   */
+  public void run(String originalQuery, String outputFile, int errorLimit) {
 
     // uses the loop function to generate and traverse the tree of possible error recoveries
-    loop(originalQuery, errorLimit);
+    // this will set the variable solution
+    loop(originalQuery, errorLimit, root, 0);
+
+    // case where entire query can be parsed
+    if (solution.equals(root)) {
+      System.out.println("The entire query can be parsed without error");
+    }
 
     // write termination logic for output (tracing the node back, reconstructing path, output)
+    Node current = solution;
+    while (current.getParent() != null) {
+      // print out the result
+      System.out.println("Unparseable portion: Start Line " + current.getStartLine() +
+          ", End Line " + current.getEndLine() + ", Start Column " + current.getStartColumn() +
+          ", End Column " + current.getEndColumn() + ", " + current.getErrorHandlingType());
+
+      // if replacement
+      if (current.getErrorHandlingType().equals("Replacement")) {
+        System.out.println(": replaced " + current.getReplaceFrom() + " with " +
+            current.getReplaceTo() + "\n");
+      }
+
+      // update for loop
+      current = current.getParent();
+    }
   }
 
   /**
-   * This is where the code for the algorithm will go: essentially, there will be a loop that
-   * constantly inputs a new query after adequate error handling
+   * This is where the code for the algorithm resides: essentially, there is a loop that
+   * constantly inputs a new query after adequate error handling. The loop terminates once
+   * the parsing doesn't throw any errors, and in the case that it went through a smaller
+   * number of unparseable components than the global minimum, it sets the solution as
+   * the global solution and also alters the minimumUnparseableComp variable.
+   *
+   * TODO: implement errorLimit logic, deal with exception casting
    */
-  private static void loop(String inputQuery, int errorLimit) {
+  private void loop(String inputQuery, int errorLimit, Node parent, int depth) {
     try {
       parser.parseQuery(inputQuery);
     } catch (Exception e) {
       // generates new queries through deletion and replacement
       SqlParserPos pos = ((SqlParseException) e).getPos();
```

Comment:
Please verify that this a sqlparseexception type before casting it into one something like

if (e instanceof SqlParseException) {
....
}

better still catch SqlParseException besides catching superclass Exception.

Reply:
Ah! I mentioned above in the PR request that Exceptions will be dealt with in future PR's: for the next few PR's I'm focusing on the main features. Will refactor the exceptions in the future, leaving this comment open for now for future reference.